engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ProgramExecutionSupport.scala

```diff
@@ -469,12 +475,37 @@ object ProgramExecutionSupport {
                 .getOrElse(expressionValue.getClass)
             )
           )
-          ctx.executionService.callFunctionWithInstrument(
-            visualization.cache,
-            visualization.module,
-            visualization.callback,
-            expressionValue +: visualization.arguments: _*
-          )
+          visualization.config.expression match {
+            case Api.VisualizationExpression.Text(_, expression) =>
+              val compiler = ctx.executionService.getContext.getCompiler
+              val ensoRootNode =
+                expressionNode.getRootNode.asInstanceOf[EnsoRootNode]
+              val compilerConfig = CompilerConfig(
+                autoParallelismEnabled = false,
+                warningsEnabled        = false,
+                isStrictErrors         = true
+              )
+              val moduleContext = ModuleContext(
+                ensoRootNode.getModuleScope.getModule,
+                compilerConfig
+              )
+              val inlineContext = InlineContext(
+                moduleContext,
+                compilerConfig,
+                Some(ensoRootNode.getLocalScope),
+                Some(false)
+              )
+              val callback = compiler.runInline(expression, inlineContext).get
+              val fn       = Function.thunk(callback.getRootNode.getCallTarget, null)
+              ctx.executionService.callFunction(fn, expressionValue)
```

Comment:
The question is: how to get the right [callerInfo.getFrame()]()? The simplest solution is to evaluate the `main` function up to the point where we want to evaluate `fn` (that shall be fast as all/most of the values are cached). As soon as we get a callback that we arrived at the desired place, we need to obtain the frame and invoke [thunkExecutorNode.executeThunk(callerInfo.getFrame(), thunk, state, _)](https://github.com/enso-org/enso/blob/08c9ecb54037396a6ea26247c37b2bccb5ac1b4f/engine/runtime/src/main/java/org/enso/interpreter/node/expression/debug/EvalNode.java#L121C12-L121C96). That's the desired result of execution.